woocommerce/woocommerce-ios
# Shipping Labels: UI for selecting a payment method

Part of #4086 

## Description

For Shipping Labels M3, we will support the ability to choose a payment method when creating a shipping label. (The ability to add a new payment method will come in a future milestone.) For this milestone, you need to already have at least one payment method (credit card) set up on your account.

By default, the last four digits of your selected credit card are displayed in the Shipping Label form and you don't have to do anything to use it. However, the payment method is editable, and selecting it will take you to a new Payment Method screen.

On the Payment Method screen, you can select a payment method from the list and toggle the email receipts setting. For now, you can interact with this screen but your changes aren't saved / the Done button is not operational. (This will be added in another PR.)

## Changes

* Adds a new reusable SwiftUI component `TitleAndToggleRow` with a simple toggle and label.
* Adds the new Payment Method screen using a new SwiftUI view `ShippingLabelPaymentMethods`.
* Adds the view model `ShippingLabelPaymentMethodsViewModel`, which handles changes to the selected payment method and email receipts setting, and contains the store owner data displayed on the Payment Method screen.
* Updates `ShippingLabelFormViewController`:
   * Sets the body text on the Payment Method cell. For now, this should update to display the selected credit card details as soon as the shipping label account settings are synced.
   * Handles navigation to the new Payment Method screen.
* Updates `ShippingLabelFormViewModel`:
   * Adds a method to handle payment method value changes.
   * Adds a method to get the payment method body (to display in `ShippingLabelFormViewController`) depending on whether there is a selected payment method.
   * Updates `syncShippingLabelAccountSettings()` to handle payment method value changes based on the account settings received from the remote API.
* Adds unit tests for handling payment method value changes.

Shipping Label Form | Shipping Label Form (editable - for testing) | Payment Method screen
-|-|-
![Simulator Screen Shot - iPhone 11 Pro - 2021-05-21 at 18 27 50](https://user-images.githubusercontent.com/8658164/119176272-c9322480-ba62-11eb-80a0-55d810f57cb1.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-05-21 at 12 31 32](https://user-images.githubusercontent.com/8658164/119175732-1cf03e00-ba62-11eb-804e-96075bcbdd00.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-05-21 at 12 31 23](https://user-images.githubusercontent.com/8658164/119176281-cdf6d880-ba62-11eb-801a-b35d0aac8437.png)


## Testing

**Setup:**

* Make sure you have at least one payment method set up in your store's WP Admin under WooCommerce > Settings > Shipping > WooCommerce Shipping (`/wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings`).
* Change this line to `editable: true` to force the Payment Method cell to be editable. (This is necessary for now, since the previous cell is not yet fully implemented and you can't get to this part of the Shipping Label flow normally.)

https://github.com/woocommerce/woocommerce-ios/blob/4690bf790e2ec7c1ba52c48c694ab4e190277011/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/Shipping%20Labels/Create%20Shipping%20Label%20Form/ShippingLabelFormViewModel.swift#L347

**To test:**

1. Launch the app in debug mode (because of the feature flag).
2. Open an order detail that is eligible for the shipping label.
3. Press the button "Create Shipping Label".
4. Wait for the Payment Method cell to become editable. Confirm this cell displays your selected credit card details (the same card selected in WP Admin) and a disclosure indicator, and no Continue button.
5. Tap the Payment Method cell. Confirm it opens the new Payment Method screen.
6. Confirm all of the credit cards in your WordPress.com account at https://wordpress.com/me/purchases/payment-methods are displayed in the list, and tapping a different one changes your selection.
7. Confirm the email receipts toggle defaults to on/off based on what you have for that setting in WP Admin, and you can toggle that setting.
8. Confirm the correct store owner details are displayed (in the payment methods list footer and in the email receipts label).

## Submitter Checklist

Update release notes:
